services/graphql/src/routes/reporting.js

```diff
@@ -0,0 +1,93 @@
+const { Parser } = require('json2csv');
+const { ObjectId } = require('mongodb');
+const Reporting = require('../services/reporting');
+const {
+  publisher, deployment, adunit, advertiser,
+  order, lineitem, ad,
+} = require('../mongoose/models');
+
+const { log } = console;
+const { isArray } = Array;
+
+const asyncRoute = fn => (req, res, next) => Promise.resolve(fn(req, res, next)).catch(next);
+const retrieve = (model, $in) => model.find({ _id: { $in } }, ['name', 'size']);
+const getModelValue = (field, row, models) => {
+  try {
+    const [model, subfield] = field.split('.');
+    const found = models[`${model}s`].filter(({ _id }) => `${_id}` === `${row[`${model}Id`]}`);
+    return found[0][subfield];
+  } catch (e) {
+    log(e);
+  }
+  return null;
+};
+
+const parseRefs = async (rows, fields) => {
+  /**
+   * Ultimately this should get each distinct model, retrieve the unique
+   * models using the fields for projection.  For now just hard loop it.
+   *
+   * @todo for large data sets this probably can't be done in memory.
+   * Is there a way to send this through graph first?
+   */
+  const models = {
+    publishers: await retrieve(publisher, [...new Set(rows.map(({ publisherId }) => publisherId))]),
```

Comment:
This entire object could utilize `Promise.all` so all queries are executed in parallel. Also, projection should be used to only return the fields needed.

Reply:
how would you implement that? Promise.all won't take an object, so unless i do some black magic with Object.keys and reduce it back together. Seems like it would just add complexity to make fast queries slightly faster?

Also the projection is being done in the `retrieve` method.

Reply:
A destructured array would work great for this :)

Don't assume "fast" queries. You're doing _seven_ queries here. Even if each take 50ms each, that's now 350ms instead of 50ms with a parallel execution. That's a 7x difference.

Reply:
Example:
```js
const [
  publisher,
  deployment,
  // etc...
] = await Promise.all([
  retrieve('publisher', rows.map(({ publisherId }) => publisherId)),
  retrieve('deployment', rows.map(({ deploymentId }) => deploymentId)),
  // etc...
]);

return {
  publisher,
  deployment,
  // etc...
};
```

Then, instead of importing each model, just import the `connection` and have the `retrieve` function do `connenction.model(modelName).find()`